utility/db2pfstream/db2pfstream.1
.TH DB2PFSTREAM 1 "$Date: 2003/03/24 14:23:07 $"
.SH NAME
db2pfstream - build database view and send metadata as a pfstream
.SH SYNOPSIS
.nf
db2pfstream  db file [-n -V -v -pf pffile -sift expression]
.fi
.SH DESCRIPTION
.LP
db2pfstream is a generalized input function for a processing system
based on a concept of a pfstream.  A pfstream is a way to 
encapsulate
pieces of metadata data that could be attached to an arbitrary 
data object to define it.  This program is designed to build a 
(potentially long) stream of these pf blocks and feed the results
into an output file.  It is expected that file would normally be
a fifo to allow the output of this program to feed directly into
a secondary processing algorithm without actually taking up 
space for the pf data.  This is intended as a simple data model
for parallel processing algorithms that are data driven like
most seismic applications.  
.LP
The \fBdb\fR and \fBfile\fR parameters are required.  
db is assumed to be an antelope database and file is the output
file name (can and would normally be a fifo created with 
mkfifo).  The default assumes a parameter file exists 
in PFPATH called db2pfstream.pf that controls most of what
this program will do (see below).
.SH OPTIONS
.IP -n
Run with no output.  Normal procedure should be to run this program
with this flag set to verify all the database attributes that are
required downstream are present in the database. 
.IP -V
Print usage line and exit immediately doing nothing. 
.IP -v
Be verbose in output.  This is most useful if the -n flag is set
but it will create a lot of messages that may not matter.  The 
primary difference is that when -v is set passthrough parameters
(see parameter file description below) are checked for null values 
and all nulls are listed. 
.IP -pf
Read from alternative parameter file pf instead of default.
.IP -sift
Apply a final subset condition \fIexpression\fR to the working
database view before starting.  This is most useful if a processing
stream is being driven by a scripting language to do something 
like event based processing one event at a time or for manually
spitting up flow between banks of processors.  
Otherwise this can generally be done more easily by modification of
the parameter file to set the input view.
.SH FILES
.LP
Here is an example pfstream output file with 5 rows and 2 attributes
(columns).  This example has no grouping defined.
.nf
example &Arr{
    ensemble    &Arr{
        00000   &Arr{
            arid        2378
            orid        316
        }
        00001   &Arr{
            arid        2379
            orid        316
        }
        00002   &Arr{
            arid        2380
            orid        316
        }
        00003   &Arr{
            arid        2381
            orid        316
        }
        00004   &Arr{
            arid        2382
            orid        316
        }
        00005   &Arr{
            arid        2383
            orid        316
        }
    }
    ensemble_keys       &Tbl{
        gridid
    }
}
__EOF__
__EOI__

For more details see pfstream(5).

.fi
.SH PARAMETER FILE
.LP
The input parameter file can get fairly long.  This is an abbreviated
example forming part of the commond arrival view used
in catalog preparations.  Check the standard one for a larger example:
.nf

# Example parameter file
sleep_time 30
dbprocess_list	&Tbl{
    dbopen event
    dbjoin origin
    dbjoin assoc
    dbjoin arrival
}
ensemble_keys	&Tbl{
    evid
}
ensemble_mode	true
group_keys	&Tbl{
}
passthrough	&Tbl{
    evname string evname
    origin.jdate	int	origin.jdate
    origin.nass	int	nass
    origin.ndef	int	ndef
    origin.ndp	int	ndp
    grn	int 	grn
    srn	int	srn
    etype	string 	etype
    review	string	review
    depdp	real	depdp
    dtype	string	dtype
    mb	real	mb
    mbid	int	mbid
    ms	real	ms
    msid	int	msid
    ml	real	ml
    mlid	int	mlid
    auth	string	auth
    origin.auth string origin.auth
    commid	int	commid
    origin.commid	int	origin.commid
    arrival.commid	int	arrival.commid
    algorithm string algorithm
    belief	double  belief
    assoc.delta double delta
    assoc.seaz  double seaz
    assoc.esaz  double esaz
    assoc.timeres  double timeres
    assoc.timedef  string timedef
    assoc.azdef  string azdef
    assoc.slodef  string slodef
    assoc.azres	double azres
    assoc.slores	double slores
    assoc.emares	double emares
    assoc.wgt double wgt
    assoc.vmodel string vmodel
    arrival.jdate int arrival.jdate
    iphase	string	iphase
    stassid int stassid
    chanid int chanid
    stype string stype
    azimuth double azimuth
    delaz double delaz
    slow double slow
    delslo double delslo
    ema double ema
    rect double rect
    amp double amp
    per double per
    logat double logat
    clip string clip
    fm string fm
    snr double snr
    qual string qual
    arrival.auth string arrival.auth
}
require	&Tbl{
    evid	int	evid
    orid	int	orid
    prefor	int	prefor
    origin.lat	real	origin.lat
    origin.lon	real	origin.lon
    origin.depth	real	origin.z
    origin.time	time	origin.time
    arid	int	arid
    phase	string	phase
    sta	string	sta
    chan	string	chan
    arrival.time	time	arrival.time
    deltim	real	deltim
}
virtual_table_name	test_arrival_view




.fi
.LP
\fIdbprocess_list\fR is a Tbl list that is passed directly to dbprocess
to build the working view for this program.   This is a VERY important
parameter in two ways.  First, it defines the set of joins that will
be needed to build the complete suite of attributes to be passed into
the ouput stream.  Second, it must define the sort order properly to 
provide the right grouping when ensemble output is requested (see below).
.LP
The boolean, \fIensemble_mode\fR, controls the basic output mode.
If this parameter is false, the program writes pf's in single 
object (row) mode.  That is, a block of parameters is written to 
the output stream for each row in the input database view.  
If \fIensemble_mode\fR is set true, the the output will be 
blocks of paremeters with repeating names surrounded by the 
parameter file block 
.nf
ensemble &Arr{
   ... 
}  
.fi
(see example above)
When \fIensemble_mode\fR is true this program searchs for 
two lists called 
\fIensemble_keys\fR and \fIgrouping_keys\fR. 
The \fIensemble_keys\fR parameter defines the grouping 
that defines one data object that is to be processed by
a downstream algorithm.  That is, an ensemble is the basic
unit of granularity of the algorithm that is to use these
data.  (Note the ensemble is assumed to have only one 
element/database row if ensemble_mode is false. )  That is,
it defines the outer blocking of the pfstream. 
The \fIgrouping_keys\fR is optional for an ensemble and 
defines a secondary grouping of the ensemble.   
A typical example of this might be an ensemble of 
three-component seismograms with the gather defining
the ensemble and the grouping defining the collections of
three-components for each station.  
.LP
\fIvirtual_table_name\fR is the tag assigned to the outer block
defining an ensemble.  That is, multiple ensembles with 
different tags can be embedded in a single pfstream block.
This allows a fairly general way to map from one name space
to another or one database schema to another.  For the example
described in the FILES section above this tag was set to example.
.LP
\fIsleep_time\fR sets the time the program will sleep before 
closing it's output file.  This should be set to the maximum
expected execution time for downstream processing of one 
data object.  It can be set to zero if the output is an
ordinary file.  
.SH "SEE ALSO"
.nf
pfstream2db(1), pstream(3), pfstream(5)
.fi
.SH "BUGS AND CAVEATS"
.LP
The sleep_time parameter is unquestionably a kludge.  It is a less
than elegant way to work around a problem with a fifo when downstream
processes operate in feeding mode (i.e. open, read one block, close)
from a fifo.  It is necessary because if db2pfstream closes and the
other end does not have the fifo open for read the last block of
data will be dropped.  This could probably be done better by using
bidirectional communication capabilities of named pipes described
in streamio(7).
.SH AUTHOR
Gary L. Pavlis, Indiana University (pavlis@indiana.edy)
.\" $Id: db2pfstream.1,v 1.3 2003/03/24 14:23:07 pavlis Exp $


oldglp/utility/db2pfstream/db2pfstream.c
/* This program builds a pfstream from a specified database view and
related list of attributes.  The pfstream is pushed to an output file
name that would normally be a fifo, but could also be a regular file.

Usage:  db2pfstream db file [-n -v -pf pffile -sift expression]

where 
db is input database
file is the output to send the stream to
-n dryrun, no output
-v verbose flag
-pf use alternate parameter pffile instead of default 
-sift apply additional subset condition to output view
	(used to avoid cluttering up parameter file script driven runs)

Most of the behaviour of this program is driven by the parameter file,
but I wont' document that here and leave it to the man page.

Author:  GAry Pavlis
Written:  September 2002
*/
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include "stock.h"
#include "db.h"
#include "elog.h"
#include "brttutil.h"
#include "pfstream.h"
#include "glputil.h"

void usage()
{
	cbanner("1.0",
		"db file [-n -V -v -pf pffile -sift expression]",
		"Gary L. Pavlis", "Indiana University",
		"pavlis@indiana.edu");
	exit(-1);
}

/*  This function adds a list of attributes stored in one
row of a datascope database to a pf.  What attributes are extracted
and converted is controlled by the Attribute_map list stored
in t.  

Arguments:
	db - row of database to use to extract info to create or
		update pf.
	t - Tbl containing pointers to Attribute_map objects that
		define which attributes to extract and possible renaming in pf
	pf - destination pf object.

Returns:
	The routine returns a Tbl list of attribute names that were null
	on this database row.  This is used in dryrun mode to flag database
	problems.  IMPORTANT:  a valid Tbl is always returned.  The caller
	must free this list.  A test for now errors is maxtbl()==0.

Author:  Gary L. Pavlis
Written:  September 2002
Modified:  march 2003
Used to return only an int. It now returns a Tbl described above.
*/
Tbl *db2pf(Dbptr db, Tbl *t, Pf *pf)
{
	int i;  
	char *nullvalue;
	int iattrib, inull;
	double dattrib, dnull;
	char sattrib[128];
	Attribute_map *m;
	Dbptr dbf;
	char *bad_attrib;
	Tbl *error_list=newtbl(0);

	for(i=0;i<maxtbl(t);++i)
	{
		m = (Attribute_map *)gettbl(t,i);
		dbf=dblookup(db,0,0,m->dbname,0);
		dbquery(dbf,dbNULL,&nullvalue);
		switch(m->type)
		{
		case DBREAL:
			if(dbgetv(db,0,m->dbname,&dattrib,0)==dbINVALID)
			{
				elog_die(0,"dbgetv error for %s\n",
					m->dbname);
			}
			dnull=atof(nullvalue);
			if(dattrib==dnull)
			{
				bad_attrib=strdup(m->dbname);
				pushtbl(error_list,bad_attrib);
				break;
			}
			pfput_double(pf,m->pfname,dattrib);
			break;
		case DBTIME:
			if(dbgetv(db,0,m->dbname,&dattrib,0)==dbINVALID)
			{
				elog_die(0,"dbgetv error for %s\n",
					m->dbname);
			}
			dnull=atof(nullvalue);
			if(dattrib==dnull)
			{
				bad_attrib=strdup(m->dbname);
				pushtbl(error_list,bad_attrib);
				break;
			}
			pfput_time(pf,m->pfname,dattrib);
			break;
		case DBINT:
			if(dbgetv(db,0,m->dbname,&iattrib,0)==dbINVALID)
			{
				elog_die(0,"dbgetv error for %s\n",
					m->dbname);
			}
			inull=atoi(nullvalue);
			if(iattrib==inull)
			{
				bad_attrib=strdup(m->dbname);
				pushtbl(error_list,bad_attrib);
				break;
			}
			pfput_int(pf,m->pfname,iattrib);
			break;
		case DBSTR:
		default:
			if(dbgetv(db,0,m->dbname,sattrib,0)==dbINVALID)
			{
				elog_die(0,"dbgetv error for %s\n",
					m->dbname);
			}
			/*conversion not necessary here since dbNULL
			query returns a string for null value*/
			if(!strcmp(sattrib,nullvalue))
			{
				bad_attrib=strdup(m->dbname);
				pushtbl(error_list,bad_attrib);
				break;
			}
			pfput_string(pf,m->pfname,sattrib);
		}
	}
	return(error_list);
}
void log_error_list(Tbl *t)
{
	int j;
	char line[512];
	int nt;
	strcpy(line,"Problem attributes: ");
	nt = maxtbl(t);
	for(j=0;j<nt;++j)
	{
		char *tval;
		tval = (char *)gettbl(t,j);
		strcat(line,tval);
		if(j<(nt-1))strcat(line,",");
	}
	elog_notify(0,line);
}
		
/* see top of file for description of the main program here */
	

int DB2PFS_verbose;
void main(int argc, char **argv)
{
	Dbptr db, dbv, dbge, dbgg;
	/*ensemble_mode is primary grouping, grouping_on set if secondary on*/
	int ensemble_mode,grouping_on;
	Tbl *ensemble_keys;
	Tbl *group_keys;
	int i;
	char *pfi=NULL;
	Pf *pf,*pfo_head,*pfo_total;
	Tbl *process_list;
	int dryrun=0;
	char *sift_exp=NULL;
	int sift;
/*Attributes listed in require will abort program if missing
passthrough parameters will generate an error message but not
cause the program to abort.  Both contain pointers to Attribute_map*/
	Tbl *require,*passthrough;
	Attribute_map *amap;
	FILE *fp;
	Tbl *error_list;
	char *tag;
	int sleep_time;  

	DB2PFS_verbose=0;

/*crack the command line*/
	if(argc<3) usage();
	elog_init(argc,argv);
	if(dbopen(argv[1],"r",&db)==dbINVALID)
	{
		elog_complain(0,"dbopen failed on database %s\n",argv[1]);
		usage();
	}

	for(i=3;i<argc;++i)
	{
		if(!strcmp(argv[i],"-pf"))
		{
			++i;
			if(i>argc) usage();
			pfi=argv[i];
		}
		else if(!strcmp(argv[i],"-V"))
			usage();
		else if(!strcmp(argv[i],"-v"))
			DB2PFS_verbose=1;
		else if(!strcmp(argv[i],"-n"))
			dryrun=1;
		else if(!strcmp(argv[i],"-sift"))
		{
			++i;
			if(i>argc) usage();
			sift=1;
			sift_exp=argv[i];
		}
		else
			usage();
	}
	if(!dryrun)
	{
		fp=fopen(argv[2],"r+");
		if(fp==NULL) usage();
	}
	
	if(pfi==NULL) pfi=strdup("db2pfstream");
	if(pfread(pfi,&pf))
		elog_die(0,"Error reading parameter file %s.pf\n",pfi);
	sleep_time=pfget_int(pf,"sleep_time");
	/* The output view gets a virtual table name that tags the
	overall collection of stuff.  This comes from the parameter file
	to make this program general, BUT it must be coordinated with
	the reader code. */
	tag = pfget_string(pf,"virtual_table_name");
	if(tag==NULL)elog_die(0,"Required parameter (virtual_table_name) missing from parameter file\n");
	ensemble_mode = pfget_boolean(pf,"ensemble_mode");
	if(ensemble_mode)
	{
		ensemble_keys=pfget_tbl(pf,"ensemble_keys");
		if(ensemble_keys==NULL)
			elog_die(0,
			 "ensemble_mode is on, but no grouping keys defined\nCheck parameter file\n");
		group_keys=pfget_tbl(pf,"group_keys");
		if(group_keys==NULL)
			grouping_on = 0;
		else
		{
			if(maxtbl(group_keys)>0)
				grouping_on = 1;
			else
				grouping_on = 0;
		}
	}
	if(DB2PFS_verbose && ensemble_mode)
	{
		char sm[128];
		char *key;
		strcpy(sm,"Defining ensemble with keys: ");
		for(i=0;i<maxtbl(ensemble_keys);++i)
		{
			key=gettbl(ensemble_keys,i);
			strcat(sm,key);
		}
		elog_log(0,"%s\n",sm);
		if(grouping_on)
		{
			strcpy(sm,"Grouping ensemble with keys:  ");
			for(i=0;i<maxtbl(group_keys);++i)
			{
				key = gettbl(group_keys,i);
				strcat(sm,key);
			}
			elog_log(0,"%s\n",sm);
		}
	}

	
	/*This function calls dbprocess using a tbl from pf*/
	dbv=dbform_working_view(db,pf,"dbprocess_list");
	if(dbv.record==dbINVALID)
		elog_die(0,"dbprocess failed:  check database and parameter file parameter dbprocess_list\n");
	if(sift_exp!=NULL)dbv=dbsubset(dbv,sift_exp,0);
	/*Now we form the groupings if needed */
	if(ensemble_mode)
	{
		dbge = dbgroup(dbv,ensemble_keys,"ensemble",1);
		if(dbge.record == dbINVALID)
			elog_die(0,"dbgroup with ensemble keys failed\n");
		if(grouping_on)
		{
			dbgg = dbgroup(dbv,group_keys,"groups",2);
			if(dbgg.record == dbINVALID)
				elog_die(0,"dbgroup on secondary grouping keys failed\n");
		}
	}

/*This builds the maps of which attributes will be saved */
	require = pfget_Attribute_map(pf,"require");
	if(require==NULL)
		elog_die(0,"Error in parameter file for parameter require\n");
	passthrough = pfget_Attribute_map(pf,"passthrough");
	if(passthrough==NULL)
		elog_die(0,"Error in parameter file for parameter passthrough\n");
	/* Unfortunately, we have two different loops for ensemble mode
	 and single object mode */
	if(ensemble_mode)
	{
		int number_ensembles;
		Dbptr db_bundle_1;
		Dbptr db_bundle_2;
		int is_ensemble, ie_ensemble;
		int isg, ieg;
		int number_groups;

		pfo_head=pfnew(0);
		pfput_tbl(pfo_head,"ensemble_keys",ensemble_keys);
		if(grouping_on) pfput_tbl(pfo_head,"group_keys",group_keys);
	

		dbquery(dbge,dbRECORD_COUNT,&number_ensembles);
		if(DB2PFS_verbose) elog_log(0,"database has %d ensembles\n",
					number_ensembles);
		if(grouping_on) 
		{
			dbgg.record=0;
			dbquery(dbgg,dbRECORD_COUNT,&number_groups);
			if(DB2PFS_verbose)
			  elog_log(0,"database has %d subgroups\n",
				number_groups);
		}
		for(dbge.record=0;dbge.record<number_ensembles;
					++dbge.record)
		{
			Pf_ensemble *pfe;
			int nmembers;
			char *grp_records;
			Tbl *grplist;

			dbgetv(dbge,0,"bundle",&db_bundle_1,0);
			dbget_range(db_bundle_1,&is_ensemble,&ie_ensemble);
			nmembers = ie_ensemble-is_ensemble;
			/* pfe does not need to hold the group records
			for this application so the ngroups variable is 
			passed as 0*/
			pfe=create_Pf_ensemble(nmembers,0);
			/*Now loop through and load the array of pf's
			with parameters defined by the attribute maps*/
			for(i=0,dbv.record=is_ensemble;i<nmembers;
				++i,++dbv.record)
			{
				pfe->pf[i]=pfnew(PFARR);
				error_list=db2pf(dbv,require,pfe->pf[i]);
				if(maxtbl(error_list)>0)
				{
				    if(dryrun)
				    {
					elog_log(0,"Ensemble %d at input database view record %d lacks required attributes\n",
						dbge.record,dbv.record);
					log_error_list(error_list);
				    }
				    else
				    {
					fprintf(fp,"%s\n",END_OF_DATA_SENTINEL);
					fflush(fp);
					elog_log(0,"FATAL ERROR: ensemble %d at input database view record %d lacks required attributes\nOUTPUT DATA STREAM TRUNCATED\n",
						i,dbv.record);
					log_error_list(error_list);
					exit(-1);
				    }
				}
				freetbl(error_list,free);
				error_list=db2pf(dbv,passthrough,pfe->pf[i]);
				if(DB2PFS_verbose)
				{
					elog_log(0,"Warning:  null passthrough attributes for ensemble %d at row %d\n",
						dbge.record,dbv.record);
					log_error_list(error_list);
				}
				freetbl(error_list,free);
				pfput_boolean(pfe->pf[i],"data_valid",1);
			}
			
			if(grouping_on)
			{
				grplist=newtbl(0);
				do {
				     dbgetv(dbgg,0,"bundle",&db_bundle_2,0);
				     dbget_range(db_bundle_2,&isg,&ieg);
				     grp_records=malloc(30);
				     sprintf(grp_records,"%d %d",
					isg-is_ensemble,ieg-is_ensemble-1);
				     pushtbl(grplist,grp_records);
				     ++dbgg.record;
				}
				while (ieg<ie_ensemble);
				pfput_tbl(pfo_head,"group_records",grplist);
			}

			pfo_total=build_ensemble(1,tag,pfo_head,pfe);
			free_Pf_ensemble(pfe);
			if(!dryrun) 
			{
				pfout(fp,pfo_total);
				fprintf(fp,"%s\n",ENDPF_SENTINEL);
				fflush(fp);
			}
			pffree(pfo_total);
			if(grouping_on)freetbl(grp_records,free);
		}
	}
	else
	{
	/*This is the simpler, single block mode */
		int nrecords;
		Pf *pfo;

		dbquery(dbv,dbRECORD_COUNT,&nrecords);
		pfo=pfnew(PFARR);
		for(dbv.record=0;dbv.record<nrecords;++dbv.record)
		{
			error_list=db2pf(dbv,require,pfo);
			if(maxtbl(error_list)>0)
			{
			    if(dryrun)
			    {
				elog_log(0,"Input database view at row %d lacks required parameters\n",
					dbv.record);
				log_error_list(error_list);
			    }
			    else
			    {
				fprintf(fp,"%s\n",END_OF_DATA_SENTINEL);
				fflush(fp);
				elog_log(0,"FATAL: input database view at row %d lacks required parameters\n",
					dbv.record);
				log_error_list(error_list);
				exit(-1);
			    }
			}
			freetbl(error_list,free);
			error_list=db2pf(dbv,passthrough,pfo);
			pfput_boolean(pfo,"data_valid",1);
			if(DB2PFS_verbose)
			{
				elog_log(0,"Warning:  null passthrough attributes for row %d of input database view\n",
					dbv.record);
			}
			if(!dryrun)
			{
				pfout(fp,pfo);
				fprintf(fp,"%s\n",ENDPF_SENTINEL);
			}
		}
	}
	if(!dryrun)
	{
		fprintf(fp,"%s\n",END_OF_DATA_SENTINEL);
		fflush(fp);
		sleep(sleep_time);
		fclose(fp);
	}
	exit(0);
}
